File: android/src/main/java/com/airbnb/android/react/maps/ViewChangesTracker.java
package com.airbnb.android.react.maps;

import android.os.Handler;
import android.os.Looper;

import java.util.LinkedList;

public class ViewChangesTracker {

  private static ViewChangesTracker instance;
  private final Handler handler;
  private final LinkedList<TrackableView> markers = new LinkedList<>();
  private boolean hasScheduledFrame = false;
  private final Runnable updateRunnable;
  private final long fps = 40;

  private ViewChangesTracker() {
    handler = new Handler(Looper.getMainLooper());
    updateRunnable = new Runnable() {
      @Override
      public void run() {
        update();

        if (markers.size() > 0) {
          handler.postDelayed(updateRunnable, fps);
        } else {
          hasScheduledFrame = false;
        }
      }
    };
  }

  public static ViewChangesTracker getInstance() {
    if (instance == null) {
      synchronized (ViewChangesTracker.class) {
        instance = new ViewChangesTracker();
      }
    }

    return instance;
  }

  public void addMarker(TrackableView marker) {
    markers.add(marker);

    if (!hasScheduledFrame) {
      hasScheduledFrame = true;
      handler.postDelayed(updateRunnable, fps);
    }
  }

  public void removeMarker(TrackableView marker) {
    markers.remove(marker);
  }

  public boolean containsMarker(TrackableView marker) {
    return markers.contains(marker);
  }

  private final LinkedList<TrackableView> markersToRemove = new LinkedList<>();

  public void update() {
    for (TrackableView marker : markers) {
      if (!marker.updateCustomForTracking()) {
        markersToRemove.add(marker);
      }
    }

    // Remove markers that are not active anymore
    if (markersToRemove.size() > 0) {
      markers.removeAll(markersToRemove);
      markersToRemove.clear();
    }
  }

}

File: android/src/main/java/com/mjstudio/reactnativenavermap/util/view/ViewChangesTracker.kt
package com.mjstudio.reactnativenavermap.util.view

import android.os.Handler
import android.os.Looper
import java.util.LinkedList

class ViewChangesTracker private constructor() {
  private val handler: Handler = Handler(Looper.getMainLooper())
  private val markers = LinkedList<TrackableView>()
  private var hasScheduledFrame = false
  private val fps = 40L

  private val updateRunnable =
    object : Runnable {
      override fun run() {
        update()

        if (markers.isNotEmpty()) {
          handler.postDelayed(this, fps)
        } else {
          hasScheduledFrame = false
        }
      }
    }

  companion object {
    @Volatile
    private var instance: ViewChangesTracker? = null

    fun getInstance(): ViewChangesTracker =
      instance ?: synchronized(ViewChangesTracker::class.java) {
        instance ?: ViewChangesTracker().also { instance = it }
      }
  }

  fun addMarker(marker: TrackableView) {
    markers.add(marker)

    if (!hasScheduledFrame) {
      hasScheduledFrame = true
      handler.postDelayed(updateRunnable, fps)
    }
  }

  fun removeMarker(marker: TrackableView) {
    markers.remove(marker)
  }

  fun containsMarker(marker: TrackableView): Boolean = markers.contains(marker)

  private val markersToRemove = LinkedList<TrackableView>()

  fun update() {
    for (marker in markers) {
      marker.update()
    }

    // Remove markers that are not active anymore
    if (markersToRemove.isNotEmpty()) {
      markers.removeAll(markersToRemove)
      markersToRemove.clear()
    }
  }
}
